### PR TITLE
Add DropConnect layer

### PR DESCRIPTION
**Pull request by abrad1212**
This will add what has been missing from this repo for a while. **DropConnect** 

I'm actually pretty confused why it wasn't already in the library because all it took was 
`K.dropout(inputs, self.rate, noise_shape=noise_shape) * self.rate`

### CODE REVIEW

**Comment by taehoonlee** [comment]
Thank you for the PR @abrad1212.
- DropConnect: random drop on weights
- Dropout: random drop on activations

This PR looks Dropout to me.

**Comment by abrad1212** [comment]
@taehoonlee Reading [here](https://nickcdryan.com/2017/06/13/dropconnect-implementation-in-python-and-tensorflow/) it states
> plain old dropout does the job so long as you remember to scale the weight matrix back down by keep_prob. After realizing that a connection weight matrix used for DropConnect is compatible input for the layer of neurons used in dropout, the only actual implementation difference between Dropout and DropConnect on TensorFlow is whether or not the weights in the masked matrix get scaled up (to preserve the expected sum).

**Comment by fchollet** [comment]
Indeed, the only implementation difference appears to be `* self.rate` in train phase. Note that `K.dropout` scaled the inputs prior to that. It's pretty strange because that can only result in worse models (in theory, and which appears to be confirmed in practice -- I wouldn't blindly trust that paper).

Overall, I don't believe such a layer is sufficiently widely in use or would be sufficiently useful to justification adding in to the core Keras API.

**Comment by abrad1212** [comment]
@fchollet Alright thanks for explaining.

**Comment by zacwellmer** [comment]
@abrad1212  Does this implementation still pass gradients to the masked weights?
